kv/bulk: compress in-memory buffers

dt opened the pull request:
Release note: none.

Release justification: high impact change with relatively low risk.

Fixes #73983.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77054)
<!-- Reviewable:end -->

dt committed:
kv/bulk: compress in-memory buffers

Release justification: high impact change with relatively low risk.

Release note: none.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=77054) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=77054) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: dt<br/>:x: craig[bot]<br/><hr/>**craig[bot]** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=77054) it.</sub>

adityamaru commented:
Going to give this a review today, sorry for the delay!

adityamaru commented on pkg/kv/bulk/buffering_adder.go:
```diff
@@ -192,49 +209,84 @@ func (b *BufferingAdder) Add(ctx context.Context, key roachpb.Key, value []byte)
+	// that justifies compression overhead as well as in relative terms to the
+	// size of the budget. The fewer/bigger buffers we have to merge, the cheaper
+	// the merged is, so we want to buffer as much as we can. However, while we're
+	// compressing we will hold both the un-comopressed and compressed in memory
```
nit: s/un-comopressed/un-compressed/g and s/buffing/buffering/g

adityamaru reviewed (approved):
I've done a couple of passes and I have no blocking comments. The changes in buffering_adder.go all make sense to me, as do the new interface and buffers in kv_buf.go. I'm finding it hard to grok the `compressBuffer` and `func (i *compressedBuffer) Next` but that could also just be because it's getting late here. I'm +1 on merging this to give it some bake time but if you think it'd benefit from an additional reviewer I'll let you tag someone in!

dt committed:
kv/bulk: split/scatter after range says it is full

Previously we would split-and-scatter after "filling" a range so that as
we continued to fill, we would do so to a new range that was first
scattered to a new node. We did this once we had written 48MB to one
range, as determined by not hitting a split while sending that many
bytes during a single flush of the above buffer. 48MB was an OK default
when the default range size was 64MB, but has long since been far below
what we now consider full. This was a minor annoyance (causing merges)
but calling split-and-scatter so much more than is needed has recently
become a more severe issue now that ~every scatter call actually moves a
range and blocks on that snapshot.

Even when 48MB was a reasonable fraction of the 64MB default, this was
just a rough approximation for full that failed to account for the
actual range's configured size and only counted data flushed since the
last explicit Flush().

Instead, this change replaces this "split when i've sent enough" logic
with a new approach that lets the range itself reply to each SST with
its remaining capacity and end key. The caller can then use this to
determine if there is room to send another file to that range or if it
should consider it full, and split if/when another key that would be
added to that ranges span is added to the batch.

This significantly reduces how often ingestion stops to wait on scatter
of a split span with the recently updated scatter behavior, more than
doubling observed speeds of bulk ingestion.

Release justification: high impact fix to new/modified functionality.

Release note (performance improvement): Ranges are split and rebalanced during bulk ingestion only when they become full, reducing excessive splits and subsequent merging and associated rebalancing-related performance impact.

dt committed:
kv/bulk: add a name to sst batcher for debug logging

Release justification: low-risk logging change.
Release note: none.

dt committed:
kv/bulk: only log >0 commit-waits and track total wait

Release note: none.

dt commented:
Added some commentary to compressedBuffer's `Next()` and to `compressBuffer` to explain what's going on (but also made it a little more complicated by adding key/key prefix sharing while I was at it.)

dt commented:
(putting this on ice for a little while)

dt commented:
@miretskiy @gh-casper after our chat yesterday, this got me thinking if we could perhaps re-use this for our batch wire format

miretskiy commented:
> @miretskiy @gh-casper after our chat yesterday, this got me thinking if we could perhaps re-use this for our batch wire format

Ack
